src/Tools/dotnet-monitor/CollectionRules/Actions/CollectDumpAction.cs

```diff
@@ -3,19 +3,80 @@
 using System;
+using System.Collections.Generic;
 using System.Threading;
 using System.Threading.Tasks;
 
 namespace Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Actions
 {
-    internal sealed class CollectDumpAction :
-        ICollectionRuleAction<CollectDumpOptions>
+    internal sealed class CollectDumpAction : ICollectionRuleAction<CollectDumpOptions>
     {
-        public Task<CollectionRuleActionResult> ExecuteAsync(CollectDumpOptions options, IEndpointInfo endpointInfo, CancellationToken token)
+        private readonly IDiagnosticServices _diagnosticServices;
+        private readonly IServiceProvider _serviceProvider;
+
+        public CollectDumpAction(IServiceProvider serviceProvider)
+        {
+            _serviceProvider = serviceProvider;
+            _diagnosticServices = serviceProvider.GetRequiredService<IDiagnosticServices>();
+        }
+
+        public async Task<CollectionRuleActionResult> ExecuteAsync(CollectDumpOptions options, IEndpointInfo endpointInfo, CancellationToken token)
+        {
+            DumpType dumpType = options.Type.GetValueOrDefault(CollectDumpOptionsDefaults.Type);
+            string egressProvider = options.Egress;
+
+            IProcessInfo processInfo = await _diagnosticServices.GetProcessAsync(new ProcessKey(endpointInfo.ProcessId), token);
+
+            string dumpFileName = DiagController.GenerateDumpFileName();
+
+            string dumpFilePath = "";
+
+            // Given our options validation, I believe this is probably redundant...should I remove it?
+            if (string.IsNullOrEmpty(egressProvider))
+            {
+                // Also, I would move this to Strings.resx if we do keep it, but I decided to wait for feedback before doing that.
+                throw new ArgumentException("No Egress Provider was supplied.");
+            }
+            else
+            {
+                KeyValueLogScope scope = DiagController.GetDumpScope(processInfo);
+
+                try
+                {
+                    dumpFilePath = await SendToEgress(new EgressOperation(
+                        token => _diagnosticServices.GetDump(processInfo, dumpType, token),
+                        egressProvider,
+                        dumpFileName,
+                        processInfo.EndpointInfo,
+                        ContentTypes.ApplicationOctetStream,
+                        scope), token);
+                }
+                catch (Exception ex)
+                {
+                    throw new CollectionRuleActionException(ex);
+                }
+            }
+
+            return new CollectionRuleActionResult()
+            {
+                OutputValues = new Dictionary<string, string>(StringComparer.Ordinal)
+                {
+                    { "EgressPath", dumpFilePath }
+                }
+            };
+        }
+
+        private async Task<string> SendToEgress(EgressOperation egressStreamResult, CancellationToken token)
```

Comment:
I'm not sure having this separate adds much value, so inline it with the caller.